# Integrative genomics sheds light on the immunobiology of tuberculosis disease in cattle

## Description

_Mycobacterium bovis_ causes bovine tuberculosis (bTB), an infectious disease of cattle that represents a zoonotic threat to humans. Research has shown that the peripheral blood (PB) transcriptome is perturbed during bTB disease but the genomic architecture underpinning this transcriptional response remains poorly understood. Here, we analyse PB transcriptomics data from 63 control and 60 confirmed _M. bovis_ infected animals and detect 2,592 differently expressed genes perturbing multiple immune response pathways. Leveraging imputed genome-wide SNP data, we characterise thousands of _cis_-expression quantitative trait loci (eQTLs) and show that the PB transcriptome is substantially impacted by intrapopulation genomic variation during _M. bovis_ infection. Integrating our _cis_-eQTL data with bTB susceptibility GWAS summary statistics, we perform a transcriptome-wide association study and identify 132 functionally relevant genes (including _RGS10_, _GBP4_, _TREML2_, and _RELT_) and provide important new omics data for understanding the host response to mycobacterial infections that cause tuberculosis in mammals.

## Overview of study

![image](https://github.com/user-attachments/assets/982b3037-6396-4eb6-812d-b1a3fb1b1ea0)


## Graph of pipeline

![image](https://github.com/user-attachments/assets/63f257fe-d8b0-4c87-8956-9b2a35f4beb1)


## Data availability

The RNA-seq data from the 60 M. bovis-infected (bTB+) and 63 control (bTB−) cattle is available at the Gene Expression Omnibus (GEO) under the BioProject Accession GSE255724. The raw UMD3.1 high density genotype data, ARS-UCD1.2 imputed and filtered data, cis-eQTL results for the three groups (bTB−, bTB+ and AAG), GWAS summary statistics for the four breed cohorts (CH, LM, HF, MB) in addition to expression models generated for use in the TWAS are available at Zenodo (https://doi.org/10.5281/zenodo.13752056). Of note, raw GWAS data were not analysed in this study; The original GWAS information can be obtained from the Irish Cattle Breeding Federation (ICBF) and additional information about sequence and genotype data availability is provided by Ring et al. Cattle GTEx Consortium cis-eQTL summary statistics can be obtained from https://cgtex.roslin.ed.ac.uk/wp-content/plugins/cgtex/static/rawdata/Full_summary_statisitcs_cis_eQTLs_FarmGTEx_cattle_V0.tar.gz.


## Support

J.F.O’G was supported by Research Ireland through the Research Ireland Centre for Research Training in Genomics Data Science (**grant no. 18/CRT/6214**). This study was also supported by Science Foundation Ireland (SFI) Investigator Programme Awards to D.E.M. and S.V.G. (**grant nos. SFI/08/IN.1/B2038 and SFI/15/IA/3154**), the University College Dublin – University of Edinburgh Strategic Partnership in One Health awarded to D.E.M., S.V.G., J.G.D.P., and E.L.C., and a Department of Agriculture, Food and the Marine (DAFM) project award to D.E.M. (**TARGET-TB; grant no. 17/RD/US-ROI/52**).

