# Bovine-eQTL-TWAS study

Mycobacterium bovis causes bovine tuberculosis (bTB), an infectious disease of cattle that represents a zoonotic threat to humans. Research has shown that the peripheral blood (PB) transcriptome is perturbed during bTB disease but the genomic architecture underpinning this transcriptional response remains poorly understood. Here, we analyse PB transcriptomics data from 63 control and 60 confirmed M. bovis infected animals and detect 2,592 differently expressed genes perturbing multiple immune response pathways. Leveraging imputed genome-wide SNP data, we characterise thousands of cis-expression quantitative trait loci (eQTLs) and show that the PB transcriptome is substantially impacted by intrapopulation genomic variation during M. bovis infection. Integrating our cis-eQTL data with bTB susceptibility GWAS summary statistics, we perform a transcriptome-wide association study and identify 132 functionally relevant genes (including RGS10, GBP4, TREML2, and RELT) and provide important new omics data for understanding the host response to mycobacterial infections that cause tuberculosis in mammals.

## Overview of study

![image](https://github.com/user-attachments/assets/982b3037-6396-4eb6-812d-b1a3fb1b1ea0)


## Graph of pipeline

![image](https://github.com/user-attachments/assets/63f257fe-d8b0-4c87-8956-9b2a35f4beb1)

 ## Data availability
 
The RNA-seq data from the 60 M. bovis-infected (bTB+) and 63 control (bTB−) cattle is available at the Gene Expression Omnibus (GEO) with the BioProject Accession GSE255724. The raw UMD 3.1 high density genotype data, ARS UCD1.2 imputed and filtered data, cis-eQTL results for the three groups (bTB−, bTB+ and AAG), GWAS summary statistics for the four breed cohorts (CH, LM, HF, MB) in addition to expression models generated for use in the TWAS are available at Zenodo (link; XXX). Of note, raw GWAS data were not analysed in this study; The original GWAS information can be obtained from the Irish Cattle Breeding Federation (ICBF) and additional information about sequence and genotype data availability is provided by Ring et al. (2019) doi: https://doi.org/10.1371/journal.pone.0212067. Cattle GTEx Consortium cis-eQTL summary statistics can be obtained from https://cgtex.roslin.ed.ac.uk/wp-content/plugins/cgtex/static/rawdata/Full_summary_statisitcs_cis_eQTLs_FarmGTEx_cattle_V0.tar.gz.

  
